components/suggest/src/yelp.rs

```diff
@@ -0,0 +1,368 @@
+        |row| -> Result<_> {
+            Ok((
+                row.get::<_, usize>("sign_index")?,
+                row.get::<_, usize>("sign_length")?,
+                row.get::<_, i8>("need_location")?,
+            ))
+        },
+    )?;
+
+    let (sign_index, sign_length, need_location) = if let Some(res) = results.pop() {
+        res
+    } else {
+        return Ok((query.trim().to_string(), None, None, false));
+    };
+
+    let pre_location = query_with_spaces
+        .get(..sign_index)
+        .map(str::trim)
+        .map(str::to_string)
+        .unwrap_or_default();
+    let location_sign = query_with_spaces
+        .get(sign_index..sign_index + sign_length)
+        .map(str::trim)
+        .filter(|s| !s.is_empty())
+        .map(str::to_string);
+    let location = query_with_spaces
+        .get(sign_index + sign_length..)
+        .map(str::trim)
+        .filter(|s| !s.is_empty())
+        .map(str::to_string);
+
+    Ok((pre_location, location_sign, location, need_location == 1))
+}
+
+/// Find the pre/post modifier from the given query string.
+/// It returns the modifiers tuple as follows:
+/// (
+///   String: Query string that is removed found the modifiers.
+///   Option<String>: Pre-modifier found in the yelp_modifiers table. If not found, returns None.
+///   Option<String>: Post-modifier found in the yelp_modifiers table. If not found, returns None.
+/// )
+fn find_modifiers(
+    conn: &Connection,
+    query: &str,
+) -> Result<(String, Option<String>, Option<String>)> {
+    if !query.contains(' ') {
+        return Ok((query.to_string(), None, None));
+    }
+
+    let words: Vec<_> = query.split_whitespace().collect();
+
+    let mut pre_modifier = None;
+    for n in (1..=MAX_MODIFIER_WORDS_NUMBER).rev() {
+        let mut candidate_chunks = words.chunks(n);
+        let candidate = candidate_chunks.next().unwrap_or(&[""]).join(" ");
+        if is_modifier(conn, &candidate, Modifier::PreModifier)? {
+            pre_modifier = Some(candidate);
+            break;
+        }
+    }
+
+    let mut post_modifier = None;
+    for n in (1..=MAX_MODIFIER_WORDS_NUMBER).rev() {
+        let mut candidate_chunks = words.rchunks(n);
+        let candidate = candidate_chunks.next().unwrap_or(&[""]).join(" ");
+        if is_modifier(conn, &candidate, Modifier::PostModifier)? {
+            post_modifier = Some(candidate);
+            break;
+        }
+    }
+
+    let mut subject_candidate = query;
+    if let Some(ref modifier) = pre_modifier {
+        subject_candidate = &subject_candidate[modifier.len()..];
+    }
+    if let Some(ref modifier) = post_modifier {
+        subject_candidate = &subject_candidate[..subject_candidate.len() - modifier.len()];
+    }
+
+    Ok((
+        subject_candidate.trim().to_string(),
+        pre_modifier,
+        post_modifier,
+    ))
+}
+
+fn is_modifier(conn: &Connection, word: &str, modifier_type: Modifier) -> Result<bool> {
+    let result = conn.query_row_and_then(
```

Comment:
```suggestion
    let result = conn.query_row_and_then_cachable(
```

(And `true` for the extra last argument, to cache it 😊)